### PR TITLE
Fixes #21938 - add plugable upgrade rake task

### PR DESCRIPTION
**Pull request by jlsherrill**
This commit adds the ability for plugins to define upgrade
rake tasks within the plugin api.  The installer can then run
the upgrade:run rake task to run them.  Tasks can be defined
to run once or to always run when the upgrade:run master task is
run.

### CODE REVIEW

**Comment by jlsherrill** [comment]
@ekohl finally :)

**Comment by jlsherrill** [comment]
TODO:
- [ ]  - Write tests once general approval is given

Example usage from plugin.rb:

```
  upgrade_task('katello:import_subscriptions', :long_running => true)
  upgrade_task('katello:upgrades:3.4:remove_unused_products', :always_run => true)
```



example output:
```
[vagrant@dev foreman]$ rake upgrade:run
/home/vagrant/git/foreman/lib/foreman.rb:8: warning: already initialized constant Foreman::UUID_REGEXP
/home/vagrant/git/foreman/lib/foreman.rb:8: warning: previous definition of UUID_REGEXP was here
API controllers newer than Apipie cache! Run apipie:cache rake task to regenerate cache.
=============================================
Upgrade Step 1/2: katello:import_subscriptions, This may take a long while.
Importing Subscriptions
=============================================
Upgrade Step 2/2: katello:upgrades:3.4:remove_unused_products
```

File: lib/tasks/upgrade.rake
Comment (ekohl) [inline lib/tasks/upgrade.rake]: Should they be sorted too or how is the order determined?

Comment (jlsherrill) [reply lib/tasks/upgrade.rake]: right now they are sorted by:
1.  plugin order
2.  order they are defined in each plugin

I think 2) generally should cover everything within a plugin, but i could see 1) causing a problem since we don't have much control over that.  If you needed to run some task before a task in another plugin there is no mechanism to do that.  I don't expect that to actually be needed, and i think we could tackle that complication if we ever did come to it.

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [inline lib/tasks/upgrade.rake]: One thing missing is the ability to pass argument of some sort to these tasks.  Some of the katello ones do use a COMMIT=true environment variable as arguments.  Couple of options:

1.  Continue using ENV variables and adjust the ENV variable prior to calling those tasks (the env variables would be defined in the plugin api)
2.  Use rake task arguments.  I really don't like these because they are only positional and not named, but it has advantages in that its not messing with the ENV variable.  

Thoughts?

Comment (ekohl) [reply lib/tasks/upgrade.rake]: We could standardize on exporting an env variable (`FOREMAN_UPGRADE=1` for example) so rake tasks can test for that.

**Review by dLobatog — COMMENTED** [review]
This looks great already @jlsherrill :+1: - is there anything you want to use it for right now?

**Comment by jlsherrill** [comment]
@ekohl added that ENV variable as well as a 'skip_failure' option to skip tasks that fail instead of stopping. 
Also added the ability to specify a rake task name different from the actual name used for tracking.  For example i could define:
```
  upgrade_task('katello::upgrade::3.6::import_subscriptions', :rake_task_name => 'katello:import_subscriptions')
```
and then later after the upgrade could define:
```
  upgrade_task('katello::upgrade::3.8::import_subscriptions', :rake_task_name => 'katello:import_subscriptions')
```
and it would track that one separately. 

 

@dLobatog yes, the goal is to migrate all the katello upgrade tasks from https://github.com/Katello/katello-installer/blob/master/hooks/post/30-upgrade.rb#L14-L108  to this mechanism.  Prs incoming soon to update:

* katello (https://github.com/Katello/katello/pull/7504)
* katello-installer
* foreman-installer

But this one is needed first.

File: app/registries/foreman/plugin.rb
Comment (ehelms) [inline app/registries/foreman/plugin.rb]: Since UpgradeTask is a Rails model/database table should this be saving at some point?

File: lib/tasks/upgrade.rake
Comment (ehelms) [inline lib/tasks/upgrade.rake:6]: Should ew namespace this? `foreman:upgrade:run` ?

File: lib/tasks/upgrade.rake
Comment (ehelms) [inline lib/tasks/upgrade.rake]: I generally find long chains hard to read.

File: lib/tasks/upgrade.rake
Comment (ehelms) [inline lib/tasks/upgrade.rake]: I think you can simplify this to:

```
tasks.select!(&:run?)
```

File: lib/tasks/upgrade.rake
Comment (ehelms) [inline lib/tasks/upgrade.rake]: So the assumption is:

 1) If it runs successfully, a database object will exist that represents the task
 2) If an upgrade task is ran, but fails, it will not be present in the database

So a lack of a database object to represent a task means either it never ran or it failed during the run. I question whether we should be tracking the latter explicitly for debugging purposes.

File: lib/tasks/upgrade.rake
Comment (ehelms) [inline lib/tasks/upgrade.rake]: This is for wanting to mark an upgrade step as complete manually?

File: app/registries/foreman/plugin/upgrade_task.rb
Comment (ehelms) [inline app/registries/foreman/plugin/upgrade_task.rb]: Nit - empty line above here

File: app/registries/foreman/plugin/upgrade_task.rb
Comment (ehelms) [inline app/registries/foreman/plugin/upgrade_task.rb]: Would Foreman ever want to use this? Would it be better then not under the plugin namespace?

File: lib/tasks/upgrade.rake
Comment (ehelms) [inline lib/tasks/upgrade.rake]: I could see this logic living in the UpgradeTask itself or at least the part in the `begin` to not require knowledge of the `rake_task_name`.

File: lib/tasks/upgrade.rake
Comment (ekohl) [reply lib/tasks/upgrade.rake]: Using `.map(&:upgrade_tasks)` would shorten it already

File: lib/tasks/upgrade.rake
Comment (ehelms) [inline lib/tasks/upgrade.rake]: Should this raise a ForemanException ?

File: app/registries/foreman/plugin/upgrade_task.rb
Comment (jlsherrill) [reply app/registries/foreman/plugin/upgrade_task.rb]: They may, and thats a good question.  I know the line has been "no new registries", and i see no where else to put this.  I could probably make it easier for foreman to define a task without defining a plugin.  Otherwise looking for guidance here from a foreman dev :)

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [reply lib/tasks/upgrade.rake]: it is yes, this may or not be needed, but its here in case we do.

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [reply lib/tasks/upgrade.rake]: what's the value in that exactly?  Logs i think would give us wayyyy more information.  When would it be useful to know that a task ran and failed, but not have any information as to why?

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [reply lib/tasks/upgrade.rake:6]: i don't think any other tasks in foreman are  namespaced?  I'm actually in favor of it, it just does not seem like the standard.

File: app/registries/foreman/plugin.rb
Comment (jlsherrill) [reply app/registries/foreman/plugin.rb]: Different upgrade task class, there's the active record model and the plugin object.   I might could rename the model to 'UpgradeTaskRecord' or something to make that more clear, open to other suggestions :)

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [reply lib/tasks/upgrade.rake]: Yeah, i kinda contemplated that, but running a rake task within a model seems incorrect.  To make it work from the model but not require it be run in a rake task, you would have to do some extra stuff like:

```
require 'rake'

Rake::Task.clear 
Foreman::Application.load_tasks
```

Rake tasks aren't really intended to be run from within the model

File: lib/tasks/upgrade.rake
Comment (ehelms) [reply lib/tasks/upgrade.rake:6]: True, but, never a bad time to start a great trend!

File: lib/tasks/upgrade.rake
Comment (ehelms) [reply lib/tasks/upgrade.rake]: Fair point, can we store why it failed? :)

Comment (ehelms) [reply lib/tasks/upgrade.rake]: Or maybe introduce an upgrade logger?

File: app/registries/foreman/plugin/upgrade_task.rb
Comment (ehelms) [reply app/registries/foreman/plugin/upgrade_task.rb]: @ekohl @tbrisker @dLobatog some thoughts here?

File: app/registries/foreman/plugin.rb
Comment (tbrisker) [reply app/registries/foreman/plugin.rb]: Please do, having two different classes with the same name is super confusing. That said, are you sure we actually need two classes? It seems to me that this separation will just add overhead and confusion about what should go where.

File: app/registries/foreman/plugin/upgrade_task.rb
Comment (tbrisker) [reply app/registries/foreman/plugin/upgrade_task.rb]: This may be useful for tasks in foreman as well, we often just push things into migrations that might not be ideal to run as such since we lack a different way to do so.

**Review by tbrisker — COMMENTED** [review]
Thinking about this, why aren't all upgrade tasks just run as part of a migration? that ensures they are executed or the system doesn't work, they are executed in a transaction so at least those that affect our db will rollback if they failed, and rails takes care of making sure they are only executed once. 
The only thing as far as I can see here that isn't addressed by migrations is the option to accept failure and printing a message for long running tasks, both of which can be handled very easily by a small modification in the migration/task. 
This means that the upgrade adds another step, of running these tasks. And when should they run? before or after migrations?
Am I missing some use case of upgrades that for some reason cannot be executed inside migrations?

**Comment by ekohl** [comment]
> Thinking about this, why aren't all upgrade tasks just run as part of a migration? that ensures they are executed or the system doesn't work, they are executed in a transaction so at least those that affect our db will rollback if they failed, and rails takes care of making sure they are only executed once.

Many upgrades in Katello talk to secondary systems like Pulp or Candlepin. Having them in database migrations means you require those systems to be present when migrating the database. This is currently true for some systems already, I'd say we should move away from this rather than to this. One could argue that the `db:migrate` + `db:seed` are just 2 more upgrade tasks we need to run rather than upgrades are part of migrations/seeding.

**Comment by tbrisker** [comment]
Ok, i can see going that way as well, it makes sense imho to try and consolidate the steps rather then add more steps. One thing this doesn't address though and migrations do handle well is the ordering. the order of the tasks isn't well defined and we might run into issues when one tasks depends on another to run earlier.

File: app/registries/foreman/plugin/upgrade_task.rb
Comment (jlsherrill) [reply app/registries/foreman/plugin/upgrade_task.rb]: @tbrisker I was kinda looking for a suggestion of how we would accomplish this, I could:

1.  Provide some easy way for foreman to add to this registry (seems kinda strange, since this a plugin api)
2.  Update the AR model to store the name, rake task, options and whether it has been run or not and rely on seeds to populate these values
3.  Create a new registry (I didn't see an existing that would fit both plugin and core together).

I would normally say 2), but that makes defining order a little weird.  The nice part of using the plugin api, is that you the order is set based upon how they are defined by the plugin.

Maybe this means 3), but I know in the past this has been frowned upon.

**Comment by jlsherrill** [comment]
> One thing this doesn't address though and migrations do handle well is the ordering. the order of the tasks isn't well defined and we might run into issues when one tasks depends on another to run earlier.

So this does handle ordering.  The tasks are executed in the order they are defined within a plugin.   Do you see it being much of an issue for cross-plugin ordering? Or a plugin's upgrade task being dependent on a foreman upgrade task?  I think that might be an easy feature to add if that ever became needed, but unless you think its likely, i think we should hold off on the added complexity.

File: lib/tasks/upgrade.rake
Comment (tbrisker) [reply lib/tasks/upgrade.rake]: `.flat_map(&:upgrade_tasks)` will shorten it even more if it in only nested 1 level.

**Comment by tbrisker** [comment]
Regarding Foreman core tasks: I think it makes sense to have this in core as well, and I would prefer having one "source of truth" regarding which tasks should run and the various attributes. Right now there is both a registry (for the plugins) and an AR (for recording which tasks ran). I guess both have their benefits and issues, but I would rather consolidate on just one of them. 
Regarding ordering: I can see a situation where we need inter-plugin or plugin-to-core dependency in upgrade tasks. There's various ways to fix this, but all of them add some complexity. I would rather we plan something that is ready for this rather than try to force it in later.

One other issue that I thought about regarding consolidation and dependencies - If we use the database even just to store which tasks have been executed already, this task can only run after `rake db:migrate` has been executed. This mean that we cannot treat `db:migrate` as just another upgrade task, but more importantly, this means that we cannot use this method for any upgrade tasks that need to run prior to migrations, if there are any such.

**Comment by jlsherrill** [comment]
@tbrisker had you given any thought to a method of specifying ordering?  I could see:

1)  allow the user to specify a simple integer ('100', '200') that tasks would be sorted by.  There would likely be some default value (100), and then we would sort by value, created_at
2) allow the user to specify a 'depends_on' some other task, else sort by created_at

I can see pros/cons of both, but curious your thoughts :)

**Comment by tbrisker** [comment]
@jlsherrill The ideas I had were both these two options or a timestamp one like migrations, but like you said none of them are ideal

**Comment by jlsherrill** [comment]
Updated to be pure AR based,populated with seeds, here's a seed example:
```
UpgradeTask.define_tasks(:katello) do
  [
    {:name => 'katello:correct_repositories', :long_running => true, :skip_failure => true, :always_run => true, :ordering => 100},
    {:name => 'katello:correct_puppet_environments', :long_running => true, :skip_failure => true, :always_run => true, :ordering => 150},
    {:name => 'katello:upgrades:3.7:make_all_ks_repos_bootable', :long_running => true}
  ]
end
```

The purpose of this define tasks method is so that tasks get updated if some option changes and any old tasks that may have been seeded previously (but not run), will be removed from the db.

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [reply lib/tasks/upgrade.rake]: That may hide some of the underlying error, not sure it would help.

**Comment by jlsherrill** [comment]
[test katello]

**Comment by akofink** [comment]
I like the integer ordering. It reminds me of configs of the format `conf.d/N-foo.conf` like we use in [forklift](https://github.com/theforeman/forklift/tree/master/vagrant/boxes.d).

**Comment by ekohl** [comment]
@akofink that pattern is copied from `run-parts(8)`

File: lib/tasks/upgrade.rake
Comment (akofink) [inline lib/tasks/upgrade.rake]: You know, this is very similar to what we were discussing a couple days ago. IMO, this should use the simplified form:

```rb
def run_task(task)
  Rake::Task[task.task_name].execute
  true
rescue => e
  ...
end
```

I'm not saying you should change it, just that there's a better way ;)

**Review by akofink — DISMISSED** [review]
Works as expected.

**Comment by jlsherrill** [comment]
@tbrisker i know you've been busy, any chance to get back to this soon?

**Comment by jlsherrill** [comment]
@tbrisker @ekohl anyway this PR could get some love?

**Comment by bbuckingham** [comment]
@tbrisker, @ekohl, If you have a moment, can you take a look at this one?  It has been sitting for a while and is needed by other PRs that are in the queue.  Thanks!

**Comment by omkarkhatavkar** [comment]
Whenever we update the recurring logic the new recurring logic will be created (new resource) and older one cancelled (not sure it is an intended behaviour, why not update the same resource? If not so we can delete that cancelled recurring as this is not of any use (as no way we can enable that) 

Any thoughts?

File: lib/tasks/upgrade.rake
Comment (ares) [reply lib/tasks/upgrade.rake]: to avoid loosing e's backtrace, Foreman::Logging.exception("...", e) should be used

File: db/migrate/20180705191153_add_upgrade_task.rb
Comment (ares) [inline db/migrate/20180705191153_add_upgrade_task.rb:4]: I'm confused, what's name, task_name and subject and how they differ? Based on what I see in tests, subject seems as foreman, katello etc. If we enforce uniqueness, why we have subject too? Just trying to understand the purpose, I think it can stay as is.

**Review by ares — DISMISSED** [review]
Looks reasonable to me, some nits found. One generic but important question I think - is it safe assumption that every upgrade task can be run while DB is operational and rails environment can be loaded? I thought that was the original reason why we had that in installer. E.g. what if we need to do some upgrade of database for which the service must be down? Wasn't foreman-maintain the desired tool to do the job? Or these are different upgrade tasks?

File: lib/tasks/upgrade.rake
Comment (tbrisker) [inline lib/tasks/upgrade.rake]: I don't think we have a localization context when running rake tasks, so no need for translations. (i may be wrong about this but afaict all other tasks don't use translated strings)

File: app/models/upgrade_task.rb
Comment (tbrisker) [inline app/models/upgrade_task.rb]: do we really need both?

File: app/models/upgrade_task.rb
Comment (tbrisker) [inline app/models/upgrade_task.rb:19]: Looks like this function deletes all existing tasks that were removed, updates all those that weren't with the same attributes and creates new ones. Wouldn't it be simpler to do something like:
```ruby
UpdateTask.where(subject: subject).destroy_all
seeded_tasks = yield
seeded_task.each do |task|
  self.create!(task.merge!(subject: subject))
end
```

File: lib/tasks/upgrade.rake
Comment (tbrisker) [inline lib/tasks/upgrade.rake:13]: What is this used for?

**Review by tbrisker — CHANGES_REQUESTED** [review]

**Comment by tbrisker** [comment]
Also I agree with @ares - Wouldn't foreman-maintain be a better place to handle upgrade tasks? That way tasks that need to run without a fully loaded DB can also be handled in the same way

File: db/migrate/20180705191153_add_upgrade_task.rb
Comment (jlsherrill) [reply db/migrate/20180705191153_add_upgrade_task.rb:4]: Name and task_name can be different because I may want to run some rake task when upgrading from version  A to version B and then again run the same task from version C to version D.   Otherwise we would have to create some new rake task just to make the 'task_name' value unique.

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [reply lib/tasks/upgrade.rake]: makes sense

Comment (jlsherrill) [reply lib/tasks/upgrade.rake]: makes sense :+1:

File: app/models/upgrade_task.rb
Comment (jlsherrill) [reply app/models/upgrade_task.rb]: see my explanation above, for marek's question

File: lib/tasks/upgrade.rake
Comment (jlsherrill) [reply lib/tasks/upgrade.rake:13]: This allows rake tasks to know if they are running as part of foreman-maintain.   A lot of the katello rake tasks do not make any changes by default (as they may be run manually), but allow a COMMIT=true option to actually commit the changes.  Instead of allowing specific options to be passed to specific rake tasks we, we ended up setting a specific ENV variable here so tasks can know that they are being run within this upgrade process and not by a user.  There was discussion around this above.

**Comment by jlsherrill** [comment]
Previously katello has used these rake tasks and anytime a rake task got added we had to go update the installer to run each rake task.  This required the author to remember that they had to file a 2nd PR on a 2nd git repo in order to accomplish this.  As you would imagine it was forgotten more times than not.  

If we moved this to foreman-maintain we would back to that 2 pr process and not any better than what we have today.

File: app/models/upgrade_task.rb
Comment (jlsherrill) [reply app/models/upgrade_task.rb:19]: If you delete them all, you no longer know which ones have been run or not.

**Comment by jlsherrill** [comment]
``` is it safe assumption that every upgrade task can be run while DB is operational and rails environment can be loaded?```

So far yes.  Generally in katello these tasks have been to manage when there is some data in the db that is related to data in a backend service.  So we need a full DB to be operational. 

I could see foreman core using this to reimport puppet classes or environments if it was ever needed.  @ekohl might be able to think of some more uses.

**Comment by ares** [comment]
So what will be the rule for deciding whether upgrade task goes here or foreman-maintain? The later has already tooling for checking whether running the task is needed, support for optional tasks, is released async etc. What is the benefit of this rake task vs foreman-maintain? Is it the use of Rails env? Cc @mbacovsky @inecas

**Comment by iNecas** [comment]
I think it makes sense to for upgrade tasks close to the Rails code-base to be defined inside the foreman/plugin code-base: foreman-maintain there can be the thing to ensure this things were run (if that would not be handled in the installer).

Also, +1 for having separate mechanism for data migration outside of the db:migrate: the db:migrate should be limited to schema changes only, but any more complex data manipulation should IMO be done after everything is migrated in the schema. I've seen very bad things happening when trying to do more complex stuff there.

What I would like to see in this PR is also ability to unit-test the upgrade code: this is another thing we are lacking in the db:migrate.

**Comment by iNecas** [comment]
Btw. just in case anyone wondered, if the keeping everything in `db:migrate`, here is the tracker issue https://projects.theforeman.org/issues/15866 I've collected some time ago, that describes why it's not a very good idea to do db transformations there. I think this PR is the right direction.

Ideally, I think the db migrations should have 3 phases:

1. `db:migrate` - for non-descructive schema changes
2. `upgrade:run` - for data transformations
3. `db:migrate:cleanup`  - for descructive schema changes (after making sure we transform the data to the new structure in step 2.

I definitely don't insist on doing all this in this PR, just laying down a possible direction of making the migrations better than we have today.

**Comment by iNecas** [comment]
Regarding the question on wether to use this or foreman-maintain for migration task, I would decide based on this decision tree:

1. does is need to happen before the rails code gets updated?
   - use foreman-maintain
2. does is need to interact with external services directly (such as migrating the qpid data scructure etc.)?
   - use foreman-maintain
3. otherwise
   - use rake upgrade:run

File: app/models/upgrade_task.rb
Comment (ekohl) [inline app/models/upgrade_task.rb]: Nit: can be `seeded_tasks.map(&:name)`

File: app/models/upgrade_task.rb
Comment (ekohl) [inline app/models/upgrade_task.rb]: Wouldn't `ensure_name` be a better name?

File: lib/tasks/upgrade.rake
Comment (ekohl) [inline lib/tasks/upgrade.rake]: This else is not needed because the if block has a return.

**Review by ekohl — COMMENTED** [review]
Should the upgrade task check if `Setting['db_pending_migration']` and `Setting['db_pending_seed']` are both false before allowing to run?

**Comment by jlsherrill** [comment]
`What I would like to see in this PR is also ability to unit-test the upgrade code: this is another thing we are lacking in the db:migrate.`

The nice part of this is that its just executing rake tasks, which can all be easily tested.  @iNecas was there something more you wanted tested?

@ares Yes, these upgrade tasks would tend to involve the rails environment and db data, but would NOT modify the structure of the rails data (since that would be left to db:migrate).   I imagine something would go in foreman-maintain if it were just doing some completely external change that requires no knowledge about rails application data.

File: app/models/upgrade_task.rb
Comment (tbrisker) [reply app/models/upgrade_task.rb:19]: ah, good point.

**Comment by iNecas** [comment]
@jlsherrill actually that's probablye enough: I was typing too fast :)

**Comment by ares** [comment]
Ok, thanks for answers. I think that it's reasonable approach, it would be good to inform dev commumity about this and ideally also document the guideline somewhere once it's merged. If there's no better place, rake task comment on top would be nice. Otherwise +1 on the direction.

**Comment by jlsherrill** [comment]
[test foreman] 
[test katello]

File: app/models/upgrade_task.rb
Comment (jlsherrill) [reply app/models/upgrade_task.rb]: actually this doesn't work.  I think you mean seed_tasks.pluck(:name)

**Comment by jlsherrill** [comment]
@tbrisker @ekohl   all updated

**Review by tbrisker — APPROVED** [review]
Works for me.

**Review by ekohl — APPROVED** [review]
Untested, but the general flow looks correct to me.

**Pull request merged**
